back/app/services/similar_ideas_service.rb

```diff
@@ -0,0 +1,53 @@
+class SimilarIdeasService
+  DEFAULT_NUM_SIMILAR_IDEAS = 10
+  DEFAULT_EMBEDDED_ATTRIBUTES = 'title_body'
+
+  attr_reader :idea
+
+  def initialize(idea)
+    @idea = idea
+  end
+
+  def similar_ideas(scope: nil, limit: DEFAULT_NUM_SIMILAR_IDEAS, embedded_attributes: DEFAULT_EMBEDDED_ATTRIBUTES, distance_threshold: nil)
+    embedding = idea.embeddings_similarities.where(embedded_attributes:).first
+    return (scope || Idea).none if !embedding
+
+    # embedding.nearest_neighbors(:embedding, distance: 'cosine') does not support
+    # applying a threshold on the neighbor_distance.
+    embeddings_pgvector = "[#{embedding.embedding.join(',')}]"
+    similarities = EmbeddingsSimilarity
+      .where.not(embedding: nil)
+      .order(ActiveRecord::Base.sanitize_sql_for_order(Arel.sql("\"embedding\" <=> '#{embeddings_pgvector}'")))
+    similarities = similarities.where('"embedding" <=> :embeddings_pgvector < :distance_threshold', embeddings_pgvector:, distance_threshold:) if distance_threshold
+
+    similarities = similarities
+      .where.not(embeddable_id: idea.id)
+      .where(embedded_attributes:)
+    similarities = similarities.where(embeddable: scope) if scope
+    similarities = similarities.limit(limit) if limit
+
+    ids = similarities.map(&:embeddable_id)
```

Comment:
I think using `:pluck(:embeddable_id)` instead of map will avoid transferring all embedding vectors, which can be significant.